File: app/csharp/AccountingSystem.Api/Infrastructure/Web/Dtos/JournalRequest.cs
using System.ComponentModel.DataAnnotations;

namespace AccountingSystem.Infrastructure.Web.Dtos;

/// <summary>
/// 仕訳作成リクエスト DTO
/// </summary>
public record JournalRequest
{
    /// <summary>仕訳伝票番号</summary>
    [Required(ErrorMessage = "仕訳伝票番号は必須です")]
    [StringLength(20, ErrorMessage = "仕訳伝票番号は20文字以内で指定してください")]
    public required string JournalNo { get; init; }

    /// <summary>起票日（YYYY-MM-DD形式）</summary>
    [Required(ErrorMessage = "起票日は必須です")]
    public required DateOnly JournalDate { get; init; }

    /// <summary>入力日（YYYY-MM-DD形式）</summary>
    [Required(ErrorMessage = "入力日は必須です")]
    public required DateOnly InputDate { get; init; }

    /// <summary>決算仕訳フラグ（0=通常、1=決算）</summary>
    public int SettlementFlag { get; init; }

    /// <summary>単振フラグ（0=複合、1=単一）</summary>
    public int SingleEntryFlag { get; init; }

    /// <summary>仕訳伝票区分</summary>
    public int JournalType { get; init; }

    /// <summary>定期計上フラグ</summary>
    public int RecurringFlag { get; init; }

    /// <summary>社員コード</summary>
    [StringLength(10, ErrorMessage = "社員コードは10文字以内で指定してください")]
    public string? EmployeeCode { get; init; }

    /// <summary>部門コード</summary>
    [StringLength(10, ErrorMessage = "部門コードは10文字以内で指定してください")]
    public string? DepartmentCode { get; init; }

    /// <summary>赤伝フラグ（0=通常、1=赤伝）</summary>
    public int RedSlipFlag { get; init; }

    /// <summary>赤黒伝票番号</summary>
    [StringLength(20, ErrorMessage = "赤黒伝票番号は20文字以内で指定してください")]
    public string? RedBlackVoucherNo { get; init; }

    /// <summary>仕訳明細リスト</summary>
    [Required(ErrorMessage = "仕訳明細は必須です")]
    [MinLength(1, ErrorMessage = "仕訳明細は1件以上必要です")]
    public required List<JournalDetailRequest> Details { get; init; }
}

/// <summary>
/// 仕訳明細リクエスト DTO
/// </summary>
public record JournalDetailRequest
{
    /// <summary>仕訳行番号</summary>
    [Required(ErrorMessage = "仕訳行番号は必須です")]
    public required int LineNumber { get; init; }

    /// <summary>行摘要</summary>
    [Required(ErrorMessage = "行摘要は必須です")]
    [StringLength(200, ErrorMessage = "行摘要は200文字以内で指定してください")]
    public required string Description { get; init; }

    /// <summary>仕訳貸借明細リスト</summary>
    [Required(ErrorMessage = "仕訳貸借明細は必須です")]
    [MinLength(1, ErrorMessage = "仕訳貸借明細は1件以上必要です")]
    public required List<JournalDetailItemRequest> Items { get; init; }
}

/// <summary>
/// 仕訳貸借明細リクエスト DTO
/// </summary>
public record JournalDetailItemRequest
{
    /// <summary>仕訳行貸借区分（D=借方、C=貸方）</summary>
    [Required(ErrorMessage = "貸借区分は必須です")]
    [RegularExpression("^[DC]$", ErrorMessage = "貸借区分は 'D'（借方）または 'C'（貸方）で指定してください")]
    public required string DebitCreditFlag { get; init; }

    /// <summary>通貨コード</summary>
    [StringLength(3, ErrorMessage = "通貨コードは3文字以内で指定してください")]
    public string CurrencyCode { get; init; } = "JPY";

    /// <summary>為替レート</summary>
    public decimal ExchangeRate { get; init; } = 1m;

    /// <summary>部門コード</summary>
    [StringLength(10, ErrorMessage = "部門コードは10文字以内で指定してください")]
    public string? DepartmentCode { get; init; }

    /// <summary>プロジェクトコード</summary>
    [StringLength(10, ErrorMessage = "プロジェクトコードは10文字以内で指定してください")]
    public string? ProjectCode { get; init; }

    /// <summary>勘定科目コード</summary>
    [Required(ErrorMessage = "勘定科目コードは必須です")]
    [StringLength(10, ErrorMessage = "勘定科目コードは10文字以内で指定してください")]
    public required string AccountCode { get; init; }

    /// <summary>補助科目コード</summary>
    [StringLength(10, ErrorMessage = "補助科目コードは10文字以内で指定してください")]
    public string? SubAccountCode { get; init; }

    /// <summary>仕訳金額</summary>
    [Required(ErrorMessage = "仕訳金額は必須です")]
    [Range(0, double.MaxValue, ErrorMessage = "仕訳金額は0以上で指定してください")]
    public required decimal Amount { get; init; }

    /// <summary>基軸換算仕訳金額</summary>
    public decimal? BaseAmount { get; init; }

    /// <summary>消費税区分</summary>
    [StringLength(2, ErrorMessage = "消費税区分は2文字以内で指定してください")]
    public string? TaxType { get; init; }

    /// <summary>消費税率</summary>
    public int? TaxRate { get; init; }

    /// <summary>消費税計算区分</summary>
    [StringLength(1, ErrorMessage = "消費税計算区分は1文字で指定してください")]
    public string? TaxCalcType { get; init; }

    /// <summary>期日</summary>
    public DateOnly? DueDate { get; init; }

    /// <summary>資金繰フラグ</summary>
    public int CashFlowFlag { get; init; }

    /// <summary>セグメントコード</summary>
    [StringLength(10, ErrorMessage = "セグメントコードは10文字以内で指定してください")]
    public string? SegmentCode { get; init; }

    /// <summary>相手勘定科目コード</summary>
    [StringLength(10, ErrorMessage = "相手勘定科目コードは10文字以内で指定してください")]
    public string? OffsetAccountCode { get; init; }

    /// <summary>相手補助科目コード</summary>
    [StringLength(10, ErrorMessage = "相手補助科目コードは10文字以内で指定してください")]
    public string? OffsetSubAccountCode { get; init; }

    /// <summary>付箋コード</summary>
    [StringLength(10, ErrorMessage = "付箋コードは10文字以内で指定してください")]
    public string? NoteCode { get; init; }

    /// <summary>付箋内容</summary>
    [StringLength(200, ErrorMessage = "付箋内容は200文字以内で指定してください")]
    public string? NoteContent { get; init; }
}

File: app/rust/accounting-system/src/infrastructure/web/dtos/journal_request.rs
use chrono::NaiveDate;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::domain::journal::{Journal, JournalDebitCreditItem, JournalDetail};

/// 仕訳作成リクエスト
#[derive(Debug, Deserialize, Serialize, ToSchema)]
pub struct JournalRequest {
    pub journal_no: String,
    pub journal_date: NaiveDate,
    pub input_date: NaiveDate,
    pub settlement_flag: i32,
    pub single_entry_flag: i32,
    pub journal_type: i32,
    pub recurring_flag: i32,
    pub employee_code: Option<String>,
    pub department_code: Option<String>,
    pub reversal_flag: i32,
    pub reversal_journal_no: Option<String>,
    pub details: Vec<JournalDetailRequest>,
}

#[derive(Debug, Deserialize, Serialize, ToSchema)]
pub struct JournalDetailRequest {
    pub journal_no: String,
    pub line_number: i32,
    pub description: String,
    pub items: Vec<JournalDebitCreditItemRequest>,
}

#[derive(Debug, Deserialize, Serialize, ToSchema)]
pub struct JournalDebitCreditItemRequest {
    pub journal_no: String,
    pub line_number: i32,
    pub debit_credit_flag: String,
    pub currency_code: String,
    pub exchange_rate: Decimal,
    pub department_code: Option<String>,
    pub project_code: Option<String>,
    pub account_code: String,
    pub sub_account_code: Option<String>,
    pub amount: Decimal,
    pub base_amount: Decimal,
    pub tax_division: Option<String>,
    pub tax_rate: Option<i32>,
    pub tax_calculation_type: Option<String>,
    pub due_date: Option<NaiveDate>,
    pub cash_flow_flag: i32,
    pub segment_code: Option<String>,
    pub contra_account_code: Option<String>,
    pub contra_sub_account_code: Option<String>,
    pub note_code: Option<String>,
    pub note_content: Option<String>,
}

impl From<JournalRequest> for Journal {
    fn from(req: JournalRequest) -> Self {
        Journal {
            journal_no: req.journal_no,
            journal_date: req.journal_date,
            input_date: req.input_date,
            settlement_flag: req.settlement_flag,
            single_entry_flag: req.single_entry_flag,
            journal_type: req.journal_type,
            recurring_flag: req.recurring_flag,
            employee_code: req.employee_code,
            department_code: req.department_code,
            reversal_flag: req.reversal_flag,
            reversal_journal_no: req.reversal_journal_no,
            details: req.details.into_iter().map(|d| d.into()).collect(),
        }
    }
}

impl From<JournalDetailRequest> for JournalDetail {
    fn from(req: JournalDetailRequest) -> Self {
        JournalDetail {
            journal_no: req.journal_no,
            line_number: req.line_number,
            description: req.description,
            items: req.items.into_iter().map(|i| i.into()).collect(),
        }
    }
}

impl From<JournalDebitCreditItemRequest> for JournalDebitCreditItem {
    fn from(req: JournalDebitCreditItemRequest) -> Self {
        JournalDebitCreditItem {
            journal_no: req.journal_no,
            line_number: req.line_number,
            debit_credit_flag: req.debit_credit_flag,
            currency_code: req.currency_code,
            exchange_rate: req.exchange_rate,
            department_code: req.department_code,
            project_code: req.project_code,
            account_code: req.account_code,
            sub_account_code: req.sub_account_code,
            amount: req.amount,
            base_amount: req.base_amount,
            tax_division: req.tax_division,
            tax_rate: req.tax_rate,
            tax_calculation_type: req.tax_calculation_type,
            due_date: req.due_date,
            cash_flow_flag: req.cash_flow_flag,
            segment_code: req.segment_code,
            contra_account_code: req.contra_account_code,
            contra_sub_account_code: req.contra_sub_account_code,
            note_code: req.note_code,
            note_content: req.note_content,
        }
    }
}
